### ISSUE TITLE
noImplicitAny not working on variables

### ISSUE THREAD

**Issue by Paril** [microsoft/TypeScript#43798]
# Bug Report

I've been on a mission to type-check our JS codebase. I'd prefer not to move to TypeScript entirely, and to just rely on using typings files and documenting our code via JSDoc. It's mostly working great, but, I'm having trouble with the `noImplicit` family of options.

For noImplicitAny: it seems to just not work at all on variables.
For all of the noImplicit family of options while working in JS files: while they sometimes show up as underlined issues that I can hover over to get context about ("but may have a better type from inferred usage"), it doesn't report it at all in Problems (which I assume TS would be responsible for populating), which is what I'm hoping to achieve.

### 🔎 Search Terms

vscode any noimplicitany problem

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about jsconfig/noImplicit

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=0&ts=4.3.0-dev.20210423&useJavaScript=true#code/DYUwLgBAHhC8EDsCuxgG4CwAobAzJCAxmAJYD2CEYIAzmABQCGAlNgN7YC+QA)

Note that this *does* report the parameter of function "test" as an error, but in VSCode with JavaScript mode, this isn't reported at all.

### 💻 Code

With a simple JS file and noImplicitAny and friends enabled in jsconfig:
```js
//@ts-check
let x = null;

function test(a)
{
}
```

### 🙁 Actual behavior

"x" shows no error, despite being an implicit any.
"a" does show an error in PlayGround, but in VSCode is only underlined in editor and unable to find a way to list all of these issues in Problems.

### 🙂 Expected behavior

Reported as warnings/errors in Problems list.

**Comment by DanielRosenwasser** [comment]
`x` is being given a control-flow analyzed automatic type. It will permit you to assign anything, but it will use that to inform the type being read.

```ts
let x = null;
x = 100;
x // hover here
```

In cases where control flow analysis can't guarantee the type, you'll get a `noImplicitAny` error.

That said, this has always been confusing. Using `any` when writing to the type, or on its declaration, has been misleading for a lot of people.

**Comment by Paril** [comment]
@DanielRosenwasser Ah, that makes sense, okay. I agree though, I'd like to disable this feature to be honest, especially in the case.

I also wasn't sure whether to report this particular issue on vscode or here regarding noImplicitAny not working in JS mode, I honestly don't know whether it's vscode or TypeScript itself that isn't marking them as errors. Your control-flow automatic type part works fine in terms of hover, but, `function(a)` only shows like a linter underline of `a` wanting a type rather than reporting it as an error.

**Comment by Paril** [comment]
Okay, so for some reason it seems like this is now "working" on my JS files. I don't know what changed or why TS has suddenly decided to show 7006 errors properly in my JS files. I did recently update VSCode (I'm on Insiders), so maybe that has something to do with it?

![](https://i.imgur.com/0VtTMAq.png)

I'm going to close this, since I think everything is indeed Working as Intended now. No idea why, but that's the magic of programming!
